app/models/results/response_tree_builder.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+module Results
+  # Builds a response tree from a form object
+  class ResponseTreeBuilder
+    def initialize(form)
+      @form = form
+    end
+
+    def build
+      @rn = AnswerGroup.new
+      add_children(@form.root_group.children, @rn)
+      @rn
+    end
+
+    def add_children(form_node, response_node)
+      form_node.each do |q|
+        if q.class == QingGroup && q.repeatable?
+          response_node.children << AnswerGroupSet.new(questioning_id: q.id, new_rank: q.rank)
+          add_repeat_groups(q, response_node.children.last)
+        elsif q.class == QingGroup
+          response_node.children << AnswerGroup.new(questioning_id: q.id, new_rank: q.rank)
+          add_children(q.children, response_node.children.last) if q.children.present?
```

Comment:
Same as above re `last`

Reply:
Also, remember that  `present?` is the same as not blank which is the same as not nil and not empty. There is no need to check both here. The best method to use is `children.any?` I would suggest, which is the opposite of `.empty?`. 

General tip: please pay attention to the subtle and exact meaning of these boolean methods.